// @uploadcare/react-adapter/dist/react-adapter.js
const P = /* @__PURE__ */ new Set([
  "children",
  "ref",
  "style",
  "className"
]), w = (e = {}) => Object == null ? void 0 : Object.keys(e).reduce((r, s) => {
  var t, c;
  const o = e == null ? void 0 : e[s], n = `on${(c = (t = o == null ? void 0 : o.split("-")) == null ? void 0 : t.map((i) => i.charAt(0).toUpperCase() + i.slice(1))) == null ? void 0 : c.join("")}`;
  return r[n] = o, r;
}, {}), m = /* @__PURE__ */ new WeakMap(), g = (e) => {
  let r = m.get(e);
  return r === void 0 && (r = /* @__PURE__ */ new Map(), m.set(e, r)), r;
}, E = ({
  node: e,
  nameProp: r,
  valueProp: s,
  prevValueProp: o,
  event: n
}) => {
  if (n !== void 0) {
    if (s !== o) {
      const t = g(e), c = t.has(n);
      let i = t.get(n);
      s !== void 0 ? c ? i.handleEvent = s : (i = { handleEvent: s }, t.set(n, i), e.addEventListener(
        n,
        (a) => i.handleEvent(a.detail)
      )) : c && (t.delete(n), e.removeEventListener(n, i));
    }
    return;
  }
  e[r] = s, s == null && r in HTMLElement.prototype && e.removeAttribute(r);
}, M = (e, r, s) => {
  const o = {}, n = {};
  return Object.entries(e).forEach(([t, c]) => {
    P.has(t) ? o[t === "className" ? "class" : t] = c : r.has(t) || t in s.prototype ? n[t] = c : o[t] = c;
  }), { reactProps: o, customElProps: n };
}, b = ({
  react: e,
  // https://react.dev/warnings/invalid-hook-call-warning
  tag: r,
  elClass: s,
  schemaEvents: o
}) => {
  const n = w(o), t = new Set(Object.keys(n ?? {})), c = e.forwardRef(
    (i, a) => {
      const f = e.useRef(/* @__PURE__ */ new Map()), p = e.useRef(null), { reactProps: h, customElProps: d } = M(i, t, s);
      return e.useLayoutEffect(() => {
        if (p.current === null)
          return;
        const u = /* @__PURE__ */ new Map();
        for (const l in d)
          E({
            node: p.current,
            nameProp: l,
            valueProp: d[l],
            prevValueProp: f.current.get(l),
            event: n[l]
          }), f.current.delete(l), u.set(l, i[l]);
        for (const [l, v] of f.current)
          E({
            node: p.current,
            nameProp: l,
            valueProp: void 0,
            prevValueProp: v,
            event: n[l]
          });
        f.current = u;
      }), e.createElement(r ?? s.__tag, {
        ...h,
        ref: e.useCallback(
          (u) => {
            p.current = u, typeof a == "function" ? a(u) : a !== null && (a.current = u);
          },
          [a]
        )
      });
    }
  );
  return c.displayName = s.name, c;
};
export {
  b as customElementToReactComponent
};
